pkg/patterns/application/application.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	internalcontroller "sigs.k8s.io/controller-runtime/pkg/internal/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// application is a very simple Controller that embeds a Manager.
+type application struct {
+	mrg    manager.Manager
+	stopFn func() <-chan struct{}
+	ctrl   *internalcontroller.Controller
+}
+
+var noop = reconcile.Func(func(reconcile.Request) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+})
+
+var getConfig = config.GetConfig
+var newController = controller.New
+var newManager = manager.New
+var getGvk = apiutil.GVKForObject
+
+// Builder builds an Application Controller and returns the manager.Manager to start the Controller.
+type Builder struct {
+	apiType runtime.Object
+	mrg     manager.Manager
+	pred    []predicate.Predicate
+	manages []runtime.Object
+	config  *rest.Config
+	r       reconcile.Reconciler
+}
+
+// BuilderFor returns a new Builder for and BuilderFor
+func BuilderFor(apiType runtime.Object) *Builder {
+	return &Builder{apiType: apiType}
+}
+
+// Owns configures the Application Controller to respond to create / delete / update events for objects it manages
+// - e.g. creates.  apiType is an empty instance of an object matching the managed object type.
+func (b *Builder) Owns(apiType runtime.Object) *Builder {
+	b.manages = append(b.manages, apiType)
+	return b
+}
+
+// WithConfig sets the Config to use for configuring clients.  Defaults to the in-cluster config or to ~/.kube/config.
+func (b *Builder) WithConfig(config *rest.Config) *Builder {
+	b.config = config
+	return b
+}
+
+// WithManager sets the Manager to use for registering the Controller.  Defaults to a new manager.Manager.
+func (b *Builder) WithManager(m manager.Manager) *Builder {
+	b.mrg = m
+	return b
+}
+
+// WithEventFilter sets the event filters.  Defaults to the empty list.
```

Comment:
`... the event filters, to filter which create/update/delete/generic events eventually trigger reconciliations`

Reply:
`for example, filtering by a particular label selector or annotation`